feat: liquidity mining update

ncitron opened the pull request:
To test, run `npm run test-update-proposal`.

This is just about ready. The only thing missing currently is the correct emissions values. One thing I noticed that was strange is that aBUSD does not have a `scaledTotalSuopply`. Is this just because it is using an older version of the contracts? Is it safe to upgrade BUSD's aToken and variabelDebtToken implementation to the most recent version?

ncitron committed:
add executor with no aToken/debtToken updates

ncitron committed:
add tests with no aToken/debtToken upgrade

ncitron committed:
add BUSD

ncitron committed:
aToken/debtToken update

kartojal commented:
Hello @ncitron, yes BUSD aToken and debt tokens are still not updated to the latest AToken or Debt token revisions. The update of incentives or the needed update of the AToken version or the Debt tokens can only be performed via Governance proposal and voted by the Aave community. You could propose to update those tokens and the incentives at the same proposal at the community forum.

ncitron commented:
Yep, that's the current plan. ARC is linked here: https://governance.aave.com/t/arc-aave-v2-liquidity-mining-program-90-days-at-30-reduced-rate/5946

eboadom commented on contracts/proposals/IncentiveUpdateExecutor.sol:
```diff
@@ -0,0 +1,127 @@
+contract IncentiveUpdateExecutor {
+
+  address constant AAVE_TOKEN = 0x7Fc66500c84A76Ad7e9c93437bFc5Ac33E2DDaE9;
+  address constant ECO_RESERVE_ADDRESS = 0x1E506cbb6721B83B1549fa1558332381Ffa61A93;
```
Minor comment but useful for clarity @ncitron . Better a name like `CONTROLLER_ECO_RESERVE` as the called contract is the controller, not the reserve itself holding the funds, and if not it confuses a bit while on code review

eboadom commented on contracts/proposals/IncentiveUpdateExecutor.sol:
```diff
@@ -0,0 +1,127 @@
+  uint256 constant DISTRIBUTION_DURATION = 7776000;   // 90 days
+  uint256 constant DISTRIBUTION_AMOUNT = 138600 ether;
+  
+  function execute(IATokenDetailed[] memory _aTokens, IATokenDetailed[] memory _debtTokens) external {
```
@ncitron not strong position, but I'm more for not passing any parameter, and hardcoding the addresses of aTokens/debtTokens on the payload, like here https://etherscan.io/address/0xC5Dc99a421e3FB9d64c7B5BBb47a5e072370eDE8#code
In my opinion is better for transparency for others to review that the parameters are correct, as checking the calldata values on the proposal creation tx is a bit more painful.

ncitron replied:
Definitely agree. I Will hardcode these values once I have done the deployments.

eboadom commented on contracts/proposals/IncentiveUpdateExecutor.sol:
```diff
@@ -0,0 +1,127 @@
+    emissions[37] = 5000000000000000;    // vDebtBUSD
+
+    // Update aTokens
+    for (uint256 i = 0; i < _aTokens.length; i++) {
```
@ncitron I guess you can unify both loops to save some gas when fetching the `underlying`, no?

ncitron replied:
One thing that makes this difficult is that for some tokens we are not updating both aTokens and debtTokens since we are only providing incentives for one side. For example, we are only upgrading DPI's aToken since there will not be borrowing incentives. Any thoughts on a better way to do this given that the aToken and vDebtToken arrays will not be the same length? I guess I can just supply address(0) in those cases to make sure the alignment is correct, and just not update the implementation if it sees address(0).

eboadom replied:
That's true, I think your idea makes sense with `address(0)`

eboadom commented on contracts/proposals/IncentiveUpdateExecutor.sol:
```diff
@@ -0,0 +1,127 @@
+
+    address[] memory assets = new address[](38);
+
+    for (uint256 i = 0; i < reserves.length; i++) {
```
@ncitron same as previous comment but here more clear in my opinion. Better to hardcode all aToken/debtToken addresses, that way you don't spend gas fetching them using `getReserveData()`

ncitron commented:
Also should I take the gas refund from the ecosystem reserved @ernesto-usal?

ncitron committed:
update proposal contract

ncitron committed:
use correct emission rates

ncitron committed:
remove comment

ncitron committed:
update tests

eboadom commented:
@ncitron if deploying the aToken/debtToken, definitely I think it is a good idea to transfer some AAVE to compensate the deployment cost, describing it on the text of the proposal

ncitron committed:
gas optimization

ncitron commented:
Yep! I've added a 70 AAVE transfer from the eco reserver to one of our wallets which should be more than enough to cover gas. I have also added it to the proposal text.
